src/modules/cmdpal/Microsoft.CmdPal.UI.ViewModels/CommandItemViewModel.cs

```diff
@@ -0,0 +1,143 @@
+﻿// Copyright (c) Microsoft Corporation
+// The Microsoft Corporation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using CommunityToolkit.Mvvm.ComponentModel;
+using Microsoft.CmdPal.Extensions;
+using Microsoft.CmdPal.Extensions.Helpers;
+using Microsoft.CmdPal.UI.ViewModels.Models;
+
+namespace Microsoft.CmdPal.UI.ViewModels;
+
+public partial class CommandItemViewModel : ObservableObject
+{
+    private readonly ExtensionObject<ICommandItem> _commandItemModel = new(null);
+
+    // private bool _initialized;
+    public string Name { get; private set; } = string.Empty;
+
+    public string Title { get; private set; } = string.Empty;
+
+    public string Subtitle { get; private set; } = string.Empty;
+
+    public string IconUri { get; private set; } = string.Empty;
+
+    public ExtensionObject<ICommand> Command { get; } = new(null);
+
+    public List<CommandContextItemViewModel> MoreCommands { get; private set; } = [];
+
+    public bool HasMoreCommands => MoreCommands.Count > 0;
+
+    public string SecondaryCommandName => HasMoreCommands ? MoreCommands[0].Name : string.Empty;
+
+    public List<CommandContextItemViewModel> AllCommands
+    {
+        get
+        {
+            var command = _commandItemModel.Unsafe?.Command;
+            var model = new CommandContextItem(command!)
+            {
+            };
+            CommandContextItemViewModel defaultCommand = new(model)
+            {
+                Name = Name,
+                Title = Title,
+                Subtitle = Subtitle,
+                IconUri = IconUri,
+
+                // TODO this probably should just be a CommandContextItemViewModel(CommandItemViewModel) ctor, or a copy ctor or whatever
+            };
+
+            var l = new List<CommandContextItemViewModel> { defaultCommand };
+            l.AddRange(MoreCommands);
+            return l;
+        }
+    }
+
+    public CommandItemViewModel(ExtensionObject<ICommandItem> item)
+    {
+        _commandItemModel = item;
+    }
+
+    public async virtual Task InitializePropertiesAsync()
+    {
+        var t = new Task(() =>
+        {
+            try
+            {
+                Initialize();
+            }
+            catch (Exception ex)
+            {
+                System.Diagnostics.Debug.WriteLine(ex);
+            }
+        });
+        t.Start();
+        await t;
+    }
```

Comment:
I think we follow a similar pattern to the ShellViewModel and the dedicated Loading page? But we do that for each view.

So, for instance, in the ListPage today, we do this:

https://github.com/zadjii-msft/PowerToys/blob/014a340c7a828fe029e934f906745c7c1eb6a41b/src/modules/cmdpal/Microsoft.CmdPal.UI/ExtViews/ListPage.xaml.cs#L30-L38

So, here we could check a prop if `IsInitialized` on the ViewModel, if it is, great we set the ViewModel as we do there now and are done.

Otherwise, we spawn off the task and call the `Initialize` method (or make it an async relay command as well - not sure how well that would work with inheritance), like the loading page did:

https://github.com/zadjii-msft/PowerToys/blob/014a340c7a828fe029e934f906745c7c1eb6a41b/src/modules/cmdpal/Microsoft.CmdPal.UI/LoadingPage.xaml.cs#L31-L41

(Though maybe unlike the Shell scenario where it sends a navigation message to finish, we're just awaiting the Task that's calling initialize or something, we can make the `OnNavigatedTo` async.)

Then in the XAML we can just overlay a Progress Ring or wrap the ListPage content in a [SwitchPresenter](https://github.com/CommunityToolkit/Windows/blob/main/components/Primitives/samples/SwitchPresenter/SwitchPresenterLoaderSample.xaml) based on the `ViewModel` being set/loaded state. `SwitchPresenter` does give us more options later if we want a third error-state if there's a problem loading the List?

Reply:
Okay I think I get why this doesn't work:

```cs
    public ListViewModel(IListPage model)
    {
        ObservableGroup<string, ListItemViewModel> group = new(string.Empty);
        foreach (var item in model.GetItems())
        {
            ListItemViewModel viewModel = new(item);
            viewModel
                .InitializePropertiesAsync()
                .ContinueWith(t =>
                {
                    group.Add(viewModel); //<-----
                });
        }
```

The `.ContinueWith` gets called on the threadpool worker thread, not the main thread. So adding the item to the ObservableCollection there explodes, because the notification isn't on the UI thread.

But we also don't want to do the resolving on the UI thread. 

Maybe we can just have some sort of goofy UI state where a list item is just like, a blur till we `Initialize` it? (I feel like I've seen this before, in Spotify for example).

Reply:
ala: ![image](https://github.com/user-attachments/assets/bed109a3-5ecc-48c6-944f-eb52e888dc7d)

Reply:
That's the shimmer control, we have it in the WCT Labs repo from the MS Store's implementation. In my branch of yours I'm just putting a progress ring while the ListViewModel hydrates all the items, we can swap it to something fancier like this later.

Yup, at some point we'll need to pass back off to the UI thread. Depending on where this happens, we have a few options. TaskScheduler is the easiest way to do this within our ViewModels project, we just have to grab a reference of it in the UI thread where the VM is initially constructed. And then once our properties are hydrated and we want to add it to the ObservableCollection we just do:

```cs
TaskScheduler _uiScheduler; // Captured/passed in from DI or construction, etc...
...
Task.Factory.StartNew(() =>
{
    OurObservableCollection.Add(finalitem); // we may create the itemviewmodel wrapper here too, we can play around with it, but just showing the general pattern  here
}, CancellationToken.None, TaskCreationOptions.None, _uiScheduler);
```